ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/util/P2PDumpManager.java

```diff
@@ -0,0 +1,106 @@
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.statetransition.util;
+
+import java.io.IOException;
+import java.nio.file.FileAlreadyExistsException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Optional;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.tuweni.bytes.Bytes;
+import org.apache.tuweni.bytes.Bytes32;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+
+public class P2PDumpManager {
+  private static final Logger LOG = LogManager.getLogger();
+
+  private static final String GOSSIP_DECODING_ERROR_DIR = "gossip_decoding_error_messages";
+  private static final String GOSSIP_REJECTED_DIR = "rejected_gossip_messages";
+  private static final String INVALID_BLOCK_DIR = "invalid_blocks";
+
+  public static Optional<String> saveGossipMessageDecodingError(
+      final Path directory,
+      final String topic,
+      final String arrivalTimestamp,
+      final Bytes originalMessage) {
+    final String fileName = String.format("%s_%s.ssz", arrivalTimestamp, topic);
+    final String identifiers = String.format("Topic: %s", topic);
+    return saveBytesToFile(
+        directory,
+        "gossip message with decoding error",
+        GOSSIP_DECODING_ERROR_DIR,
+        fileName,
+        identifiers,
+        originalMessage);
+  }
+
+  public static Optional<String> saveGossipRejectedMessageToFile(
+      final Path directory,
+      final String topic,
+      final String arrivalTimestamp,
+      final Bytes decodedMessage) {
+    final String fileName = String.format("%s_%s.ssz", arrivalTimestamp, topic);
+    final String identifiers = String.format("Topic: %s", topic);
+    return saveBytesToFile(
+        directory,
+        "rejected gossip message",
+        GOSSIP_REJECTED_DIR,
+        fileName,
+        identifiers,
+        decodedMessage);
+  }
+
+  public static Optional<String> saveInvalidBlockToFile(
+      final Path directory, final UInt64 slot, final Bytes32 blockRoot, final Bytes blockSsz) {
+    final String fileName =
+        String.format("slot%s_root%s.ssz", slot, blockRoot.toUnprefixedHexString());
+    final String identifiers = String.format("Slot: %s, Block Root: %s", slot, blockRoot);
+    return saveBytesToFile(
+        directory, "invalid block", INVALID_BLOCK_DIR, fileName, identifiers, blockSsz);
+  }
+
+  private static Optional<String> saveBytesToFile(
+      final Path baseDirectory,
+      final String object,
+      final String errorDirectory,
+      final String fileName,
+      final String identifiers,
+      final Bytes bytes) {
+    // Check directory exists
+    final Path directory = baseDirectory.resolve(errorDirectory);
+    if (directory.toFile().mkdir()) {
+      LOG.info(String.format("%s directory has been created to save %s.", errorDirectory, object));
+    }
+    final Path path = directory.resolve(fileName);
```

Comment:
This is going to check if the directory exist every single time, this seems inefficient and we could only check it like 3 times in the worst scenario (1 for each subdirectory or something like that). Imagine getting thousands of invalid voluntary_exits in a topic, we would be doing thousand of directory exits checks.